### ISSUE TITLE
Allow defining payloads with non-JSON standards (XSD, Protobuf, etc.)

### ISSUE THREAD

**Issue by jonaslagoni** [asyncapi/spec#881]
I choose to start a new issue to map out more precisely what is needed to support this, without any specific sight on one approach or the other. This is a generic solution that would work with all non-JSON standards regardless of which one.

This was first pursued for V3 of the spec, but I don't believe that it will be [possible for that deadline](https://github.com/asyncapi/spec/issues/880). Therefore I gathered the current progress to have one place to discuss it so it can be removed [from the progress list](https://github.com/asyncapi/spec/issues/691).

Steps currently being taken:
- [ ] Figure out if we need a new referencing standard or not.
  - This is not explicitly agreed upon, but so far it seems like it's not sufficient with the current standard.
- [ ] If we do, which one should we go with?
  - Alternatives here: https://github.com/json-schema-org/referencing
    - Overall perspective: https://github.com/json-schema-org/referencing/issues/14
    - JRI perspective: https://github.com/json-schema-org/referencing/issues/17
    - JRef perspective: https://github.com/json-schema-org/referencing/issues/16
  - Custom alternative: https://github.com/asyncapi/spec/pull/825
- [ ] Implement a reference handling tool in TypeScript (this is bare minimum) that can handle multiple format types and be context dependant. 
  - Look through the requirement discussion here: https://github.com/orgs/asyncapi/discussions/485

Related resources:
- https://github.com/asyncapi/spec/issues/622
- https://github.com/asyncapi/spec/issues/624
- https://github.com/asyncapi/spec/issues/216
- https://github.com/asyncapi/spec/issues/434
- https://github.com/orgs/asyncapi/discussions/485
- https://github.com/asyncapi/spec/pull/825
- https://github.com/json-schema-org/referencing
- https://github.com/OAI/OpenAPI-Specification/issues/764
- https://github.com/OAI/OpenAPI-Specification/issues/1532

**Comment by derberg** [comment]
I looked into the linked draft PR and the long discussion there. It was hard to follow as there was a lot of OpenAPI and JSON Schema "history" involved and I got lost.

just to make it clear: **for me this is clearly only about the message payload referencing** so referencing of schemas only, not referencing other parts of AsyncAPI

what I understand is:
- the community needs a way to reference the payload of schema in xsd or proto
- I assume we will need also something for GraphQL at some point of time
- sometimes it is about referencing just a file, sometimes part of the file, as both in xsd and proto you can have some specific parts to reference too, like you can have multiple messages in one proto file
- `$ref` cannot be used as it is for JSON

How about solving it by having another way of referencing things:
- so alongside Schema Object, Reference Object we could have something like Schema Reference Object
- `ServerlessWorkflow` spec has it already (different structure, pointing them out as example of custom pointers) and it makes lots of sense -> https://github.com/serverlessworkflow/specification/blob/main/specification.md#workflow-functions and the structure is pretty clear
- long term Reference Object could be deprecated (AsyncAPI 5.0 - although I do not have strong opinion on deprecation need) and people would always use Schema Reference Object

So in short `$ref` is just for bundling multiple JSON/YAML AsyncAPI structures into one, referencing inside document different structures. It is better and more intuitive for users to reference schemas in a more defined way when you can say that whatever comes after `#` is a proto message:
```
"reference": "https://gist.githubusercontent.com/iamsurajgharat/1d43d8181f286a58617adebce1714c6f/raw/343d91d8e3f09ab717f50bffa3213a7fd8fb0e13/sumservice.proto#SumInput"
```
or xsd element:
```
"reference": "https://gist.githubusercontent.com/iamsurajgharat/1d43d8181f286a58617adebce1714c6f/raw/343d91d8e3f09ab717f50bffa3213a7fd8fb0e13/sumservice.proto#ldml"
```

the only problem I see is with the `schemaFormat`. I'm not 100% sure yet how that would play. The more I think about it, my view is that `schemaFormat` should not be defined outside `payload` as it is now.

PS. I do not worry that we will not have out-of-the-box tooling support as we have for `$ref`. We will just make sure to have current Studio and IDEs plugins support it and that should be really enough.

**Comment by jonaslagoni** [comment]
> just to make it clear: for me this is clearly only about the message payload referencing so referencing of schemas only, not referencing other parts of AsyncAPI

Agreed, anything else would just complicate things I think.

> So in short $ref is just for bundling multiple JSON/YAML AsyncAPI structures into one, referencing inside document different structures. It is better and more intuitive for users to reference schemas in a more defined way when you can say that whatever comes after # is a proto message:

That would definitely be possible yes.

Do you know if serverlessworkflow has a core tool for solving references we could maybe use if this became the case? 

> PS. I do not worry that we will not have out-of-the-box tooling support as we have for $ref. We will just make sure to have current Studio and IDEs plugins support it and that should be really enough.

That is a contradicting sentence, cause Studio and IDE plugins can't support it without the core tool in place 🤷 Unless they implement the functionality themself of course, but then it might as well be in a separate tool everyone can integrate. Or did I misunderstand you? 🤔

**Comment by jonaslagoni** [comment]
So started specific discussions for the [2 proposals in JSON Schema referencing repository](https://github.com/json-schema-org/referencing/issues/14) and answered the outstanding clarification questions in separate issues for each standard
- JRI: https://github.com/json-schema-org/referencing/issues/17
- JRef: https://github.com/json-schema-org/referencing/issues/16

They seem to both support what we want, and leave it up to the context specification (AsyncAPI) to dictate the behavior necessary for referencing non-JSON Standards.

**Comment by derberg** [comment]
> Do you know if serverlessworkflow has a core tool for solving references we could maybe use if this became the case?

I do not think they have any standard there. Their needs are also super specific, like for example in case of referencing AsyncAPI, reference points to `operationId`, so they for sure just pull out the message payload of that operation "manually" without any standard in use.

Regarding `JRI` or `JRef` -> I could not find any reference on internet, what are these standards all about? and are they useful if there are no reference implementations anyway?

**Comment by GreenRover** [comment]
+1
I am interested in protobuff support as well.

**Comment by GreenRover** [comment]
@derberg i dont see GraphQL support.
Because as i understand GraphQL it is HTTP based only. And dont know concepts like channels.
Or differently sayed: channels/operations are all parts of the GraphQL schema.
So GraphQL is more a competitor of AsyncAPI as something that could be integreated

**Comment by rober15** [comment]
I think we have as well to consider another type of async messaging added now as well to Open API:
https://swagger.io/docs/specification/callbacks/

So i wonder how to handle that Callbacks which can be used for webhooks but shown under Open API which most people associate more with synchronous messaging and not as well async.

**Comment by AnimeshKumar923** [comment]
Is it still valid? @jonaslagoni @derberg
